src/hotspot/share/runtime/crac.cpp

```diff
@@ -602,3 +656,1075 @@ void crac::update_javaTimeNanos_offset() {
+        ResourceMark rm;
+        THROW_MSG_(vmSymbols::java_lang_IllegalArgumentException(),
+                   err_msg("Object array dump " INT64_FORMAT " has illegal class %s", dump.id, k->external_name()),
+                   {});
+      }
+      klass = ObjArrayKlass::cast(k);
+    }
+
+    // Ensure won't overflow array length which is an int
+    if (dump.elem_ids.size() >  std::numeric_limits<int>::max()) {
+      THROW_MSG_(vmSymbols::java_lang_IllegalArgumentException(),
+                 err_msg("Object array dump " UINT64_FORMAT " has too many elements: " UINT32_FORMAT " > %i",
+                         dump.id, dump.elem_ids.size(),  std::numeric_limits<int>::max()), {});
+    }
+    int elems_num = static_cast<int>(dump.elem_ids.size());
+
+    objArrayHandle handle;
+    {
+      objArrayOop o = klass->allocate(elems_num, CHECK_({}));
+      handle = objArrayHandle(Thread::current(), o);
+    }
+
+    assert(!_restored_objects.contains(dump.id), "Should still be not restored");
+    _restored_objects.put_when_absent(dump.id, handle);
+
+    restore_class(*class_dump, CHECK_({}));
+
+    for (int i = 0; i < elems_num; i++) {
+      Handle elem = restore_object(dump.elem_ids[i], CHECK_({}));
+      if (elem.is_null() || elem->klass()->is_subtype_of(klass->element_klass())) {
+        handle->obj_at_put(i, elem());
+      } else {
+        ResourceMark rm;
+        THROW_MSG_(vmSymbols::java_lang_IllegalArgumentException(),
+                   err_msg("%s array has element %i of incompatible type %s in the dump",
+                           klass->element_klass()->external_name(), i, elem->klass()->external_name()),
+                   {});
+      }
+    }
+
+    if (log_is_enabled(Trace, restore)) {
+      ResourceMark rm;
+      log_trace(restore)("Restored object array " INT64_FORMAT " (%s)", dump.id, klass->external_name());
+    }
+    return handle;
+  }
+
+  typeArrayHandle restore_prim_array(const HeapDumpFormat::PrimArrayDumpRecord &dump, TRAPS) {
+    assert(!_restored_objects.contains(dump.id), "Use restore_object() which also checks for ID duplication");
+    log_trace(restore)("Restoring primitive array " INT64_FORMAT, dump.id);
+
+    // Ensure won't overflow array length which is an int
+    if (dump.elems_num > std::numeric_limits<int>::max()) {
+      THROW_MSG_(vmSymbols::java_lang_IllegalArgumentException(),
+                 err_msg("Primitive array dump " UINT64_FORMAT " has too many elements: " UINT32_FORMAT " > %i",
+                         dump.id, dump.elems_num,  std::numeric_limits<int>::max()), {});
+    }
+    int elems_num = static_cast<int>(dump.elems_num);
+
+    typeArrayOop o;
+    switch (dump.elem_type) {
+      case HPROF_BOOLEAN:
+        o = oopFactory::new_typeArray_nozero(T_BOOLEAN, elems_num, CHECK_({}));
+        precond(elems_num * sizeof(jboolean) == dump.elems_data.size());
+        if (elems_num > 0) memcpy(o->bool_at_addr(0), dump.elems_data.mem(), dump.elems_data.size());
```

Comment:
Just wanted to make sure, is doing `memcpy` into primitive array memory like this always safe?

Reply:
Yeah, seems to be, assuming the `precond()` is the `guarantee()`